### PR TITLE
Advection direction

### PR DESCRIPTION
**Pull request by ian-r-rose**
Attempt at addressing #526, as well as some conversations I had with Cedric and Gerry at the hackathon.

At the moment, we advect the free surface in the normal direction of the surface.  This involves a finite element projection of the velocity solution into the direction of the surface normal.  With this scheme the surface advection has the same mass conservation properties (more or less) as the div u = 0 block of the Stokes solve.  The downside is that for large mesh deformation this projection onto the surface normal can go unstable, as seen in #526, among other places.

An alternative scheme would be to project the velocity solution onto the local vertical direction instead of the normal direction.  Then the mesh nodes would be moved only up and down.  In fact, a previous version of this code had this scheme, but I changed it because the mass conservation properties were worse, and the global mesh volume could drift a bit over time.  The benefit of this, however, is that the mesh tends to see fewer of these "pinching" instabilities.
Here is an image of the pinching that can happen with large curvature of the surface due to a sinking blob:
normal

Clearly bad.  If I advect only in the local vertical, however, I get something like this:
gravity

Nicer.  As it stands, I have implemented a switch to toggle between these two different schemes.  I would like to test this with the crustal deformation cookbook as well, but I see that the Drucker-Prager material model has not been merged yet.  Thoughts?

### CODE REVIEW

**Comment by bangerth** [comment]
The figures clearly look different, but how much of that is due to the fact that the second model has the two big cells at the top center? How would things look like if they had the same mesh?

**Comment by ian-r-rose** [comment]
Still very different. I believe the issue comes about due to the projection onto the normal direction losing accuracy when neighboring normals are too different, so when the curvature is high you can get a runaway pinch.

This could be alleviated somewhat by going to higher resolution, but in practice, any kind of strong variation in surface velocity could produce these types of effects (changes in density, viscosity, etc).

normal

gravity

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -1434,6 +1434,20 @@ namespace aspect
...
+          struct SurfaceAdvection
+          {
+            enum Direction { normal, gravity };
+          };
+
+          /**
+           * Stores whether to advect the free surface in the normal direction
+           * or the direction of gravity.
+           */
+          typename SurfaceAdvection::Direction advection_direction;
```
Comment (bangerth) [inline include/aspect/simulator.h]: No need to use `typename` here. (In fact, it's forbidden pre-C++11.)

File: source/simulator/freesurface.cc [hunk 2]
```diff
@@ -292,20 +315,30 @@ namespace aspect
...
-                                      fs_fe_face_values.JxW(point);
-                  }
+                {
+                  //Select the direction onto which to project the velocity solution
+                  Tensor<1,dim> direction;
+                  if ( advection_direction == SurfaceAdvection::normal ) //project onto normal vector
+                    direction = fs_fe_face_values.normal_vector(point);
+                  else if ( advection_direction == SurfaceAdvection::gravity ) //project onto local gravity
+                    direction = sim.gravity_model->gravity_vector( fs_fe_face_values.quadrature_point(point) );
+                  else ExcInternalError();
```
Comment (bangerth) [inline source/simulator/freesurface.cc]: This last line creates an unnamed ExcInternalError object, destroys it again, and goes on with life :-)

You'll need something like
```
  AssertThrow (false, ExcInternalError());
```

File: source/simulator/freesurface.cc [hunk 3]
```diff
@@ -292,20 +315,30 @@ namespace aspect
...
+                  direction = direction/direction.norm();
```
Comment (bangerth) [inline source/simulator/freesurface.cc]: This led to divisions by zeros in other contexts. Write it as
```
  direction *= (direction.norm()>0 ? 1./direction.norm() : 0);
```
or similar.

File: include/aspect/simulator.h (hunk 1 above)
Comment (ian-r-rose) [inline include/aspect/simulator.h]: Hmm, compilation fails without it.  I modeled this struct after the NonlinearSolver and NullspaceRemoval structs.

**Comment by bangerth** [comment]
I'm agnostic -- you're the expert in the area :-)

Fix the minor issues and add a testcase, and let's get it committed.

File: include/aspect/simulator.h (hunk 1 above)
Comment (bangerth) [inline include/aspect/simulator.h]: That's too weird. But if it works (and apparently it does), then go with it -- I stand corrected.

**Comment by bangerth** [comment]
This has become outdated :-(  Can you rebase to current master?

**Pull request merged**